Improve error handing when JSON schema conversion fails

wbolster opened the pull request:
Use a template, improve the error message, and add tests.

tomchristie commented on tests/test_json_schema.py:
```diff
@@ -142,5 +142,9 @@ class CustomField(typesystem.Field):
         to_json_schema(field)
+
+    message = str(exc_info.value)
+    assert message.startswith("Unsupported field type")
```
Good call!

Can we just use `assert message == "Unsupported field type 'CustomField' passed to 'to_json_schema"` here?

wbolster replied:
no, because the error is ```
Unsupported field type 'test_to_json_schema_invalid_field.<locals>.CustomField' passed to 'to_json_schema'
``` in full.

wbolster replied:
no, because the error is

```
Unsupported field type 'test_to_json_schema_invalid_field.<locals>.CustomField' passed to 'to_json_schema'

```
in full.

tomchristie commented on typesystem/json_schema.py:
```diff
@@ -435,7 +435,9 @@ def to_json_schema(
     field_type = type(field)
-    raise TypeError("Unsupported field type {field_type} passed to 'to_json_schema'")
+    raise TypeError(
+        f"Unsupported field type {field_type.__qualname__!r} passed to 'to_json_schema'"
```
Great - yup. I'd prefer if we can keep attribute lookups pulled outside of `f-strings`, so perhaps...

```python
field_qualname = type(field).__qualname__
raise TypeError("Unsupported field type {field_qualname!e} passed to 'to_json_schema'")
```

Or something similar?

wbolster replied:
simplified it even more

tomchristie commented:
Yes! Neat improvement here! Couple of possible tweaks.

wbolster commented:
some changes:

- moved the custom field definition outside the test so that the `__qualname__` doesn't have `...<locals>.CustomField`
- changed the error message. no reason to have the function name there. that's what stack traces are for.
- raise `ValueError`, not `TypeError`. it's an invalid value after all. yes, the argument is a class (aka a `type` in python), but `TypeError` is typically reserved for "wrong usage of the api". in this case the api accepts a field, which is correct, but the function just doesn't know how to convert this particular one. it also aligns better with the checks in #44 which also raise `ValueError` for exactly the same reason.

tomchristie commented:
There was a merge conflict here, but I'm not sure I've resolved it correctly?

wbolster commented:
i'll take case of it

wbolster commented:
all good now!

tomchristie commented:
Ace!
